### ISSUE TITLE
Android: Add more touchscreen buttons

### ISSUE THREAD

**Issue by Pantyhose-X** [luanti-org/luanti#13229]
## Problem
Add virtual controls for easier operation

## Solutions
Custom Controls https://www.youtube.com/watch?v=pNUYuymoiuA
![image](https://user-images.githubusercontent.com/42504385/219325283-dc8f31d8-128c-41e5-b3e2-3ba851493efa.png)

![](https://preview.redd.it/xcxqwm4q2js81.jpg?width=2400&format=pjpg&auto=webp&v=enabled&s=64572b06b83c47e2609d7e549417b648bc8261df)

**Comment by Unacceptium** [comment]
would solve this:
https://forum.minetest.net/viewtopic.php?p=429136#p429136

**Comment by grorp** [comment]
I have a branch with virtual button controls here: https://github.com/grorp/minetest/tree/the-good-stuff

![screenshot](https://github.com/minetest/minetest/assets/82708541/f6e54176-9fab-4c3f-83e3-5881af12c33c)

**Comment by Unacceptium** [comment]
@grorp wheres the apks? i'm new to github

**Comment by grorp** [comment]
https://github.com/grorp/minetest/actions/runs/6318026396

**Comment by Unacceptium** [comment]
actually all makes a mixture of whats needed, like we need more buttons and ability to use most of them at once,(like aux+sneak etc) but minimal customization surely needed. virtual controls mean you place and drag the buttons or sliders anywhere you want As An User! but for first and most important is more aux buttons at all, jump and sneak buttons higher placement, left click and right click buttons....... all can be placed by player anywhere it wants!
sorry for bad english

**Comment by Unacceptium** [comment]
@grorp cannot install it says

"an application that is not installed as a package is invalid"

**Comment by Unacceptium** [comment]
XD this is minetest not minecrap! @bplx 
yes i also hope that, this fork is nearly perfect for me
except the inventory issue in 5.8

**Comment by tetsuo55** [comment]
> I have a branch with virtual button controls here: https://github.com/grorp/minetest/tree/the-good-stuff
> 
> ![screenshot](https://user-images.githubusercontent.com/82708541/271187400-f6e54176-9fab-4c3f-83e3-5881af12c33c.jpg)

Using this build when flying below the island of the end in creative mode in mineclone2 the flights controls stop working and I fall into the abyss

**Comment by Unacceptium** [comment]
@tetsuo55 what the frick? can you make a video about that?

**Comment by tetsuo55** [comment]
> @tetsuo55 what the frick? can you make a video about that?

Ok I will do it as soon as I am able to.

Also this is the only build where (cross)bow and arrow works on mineclone2

**Comment by Unacceptium** [comment]
@tetsuo55 i tried that thing and does not happen for me. i flew below end islands and everything is ok. Please mind that sneak button is not a button! its a switch/toogle! if thats the problem, just press it again!

**Comment by tetsuo55** [comment]
> @tetsuo55 i tried that thing and does not happen for me. i flew below end islands and everything is ok. Please mind that sneak button is not a button! its a switch/toogle! if thats the problem, just press it again!

You're right  , sneak being a toggle it forces you down while in flight mode until you press it again

**Comment by grorp** [comment]
> However shields still can't be used because the button for it appears under the directional controls

@tetsuo55 Have you tried disabling `virtual_joystick_triggers_aux1`?

**Comment by Unacceptium** [comment]
> > @tetsuo55 
> However shields still can't be used because the button for it appears under the directional controls
i disagree. see screenshots below:

**Comment by tetsuo55** [comment]
> > > @tetsuo55
> > > However shields still can't be used because the button for it appears under the directional controls
> 
> ![Screenshot_20231106_163748](https://user-images.githubusercontent.com/145642963/280764157-202c4539-9d9c-4c34-804f-23ff70bc6655.jpg) ![Screenshot_20231106_163727](https://user-images.githubusercontent.com/145642963/280764210-894ba9ef-9d0c-40a2-8ce8-296d4b7c1458.jpg)
> 
> i disagree. see screenshots below:
Thanks so resolution seems to play a role, on my phone the button for the shield almost touches the joystick (I can't upload a screenshot for some reason)
Can you actually block something when its in the right hand?.

**Comment by Unacceptium** [comment]
what button? click right mouse button? theres no extra buttons for shield?

**Comment by tetsuo55** [comment]
> what button? hold right mouse button? theres no extra buttons for shield?

Thanks for teaching me how to use the shield, I thought you had to press and hold the shield icon itself on the bottom left

**Comment by Unacceptium** [comment]
theres so much similarity between @grorp 's fork and Minecrap new controls...
![Screenshot_20231107_193243](https://github.com/minetest/minetest/assets/145642963/0fb91486-aada-4ccb-a265-e06cbefe3d40)
![Screenshot_20231107_193526](https://github.com/minetest/minetest/assets/145642963/6a9abe3c-2c7e-4eee-b10b-a1494d71ca20)

also, Minecrap got Virtual Controls!!
![Screenshot_20231107_191901](https://github.com/minetest/minetest/assets/145642963/d28a93c6-b42d-4392-8e69-f8703b1c1bda)

**Comment by AdityaMH** [comment]
still in development?

**Comment by grorp** [comment]
I'm moving the "allow customizing the button layout" part of this issue to #14456, since the other issue is more clear.

This one can stay open for the "add more on-screen buttons" part.

**Comment by AdityaMH** [comment]
one more annoying thing is, there no attack/interact switch button (most android need) or attack/switch button (this only for using crosshair, still hard to imagine) in Android. since pc can press physical keyboard.

**Comment by medleymedley** [comment]
Now that an editor has been implemented thanks to grorp this will probably be easier! 

I imagine the easiest option for giving the player the ability to add buttons (not already in the pause menu), is for the "Add button" prompt in #14933 to show a new "other buttons" button to the pause menu. This could open up a clickable list of the Luanti key-bindings, taken directly from the internal keybindings if possible. 

I agree with @AdityaMH, I think the option to have separate attack/interact buttons would be useful for many, which is why I think the menu customizer should be "finalized" with the option to add any PC keybinding as well. Allowing any keybinding to be added would broadly make it easier for people to customize things to their need, like with [#14388](https://github.com/luanti-org/luanti/issues/14388), by making a DPAD with the arrow buttons.
